arm-none-eabi: build error ('float_t' undeclared)

This error occurs when running `./ct-ng arm-none-eabi && ./ct-ng build`. It's the default configuration.

### Issue submission guidelines
See [detaled guidelines](http://crosstool-ng.github.io/support/). In short:
- Report crosstool-NG version: https://github.com/crosstool-ng/crosstool-ng/commit/e8b3eccd8a036e6f3602cd25b2307e3db5476734
- Report host OS and its version: Gentoo amd64 (gcc 12.2.1_p20230428-r1)
- Attach `.config` (unless the issue is not related to any configuration): [.config.gz](https://github.com/crosstool-ng/crosstool-ng/files/11471674/default.config.gz)
- Attach `build.log` (unless crosstool-NG fails before the build starts): [build.log.gz](https://github.com/crosstool-ng/crosstool-ng/files/11471672/build.log.gz)

Relevant snippets:
```
[ALL  ]    arm-none-eabi-cc -B/home/mid-kid/.local/opt/cross/crosstool-ng/.build/arm-none-eabi/build/build-libc/arm-none-eabi/thumb/v8.1-m.main+pacbti+fp/bp/softfp/newlib/ -isystem /home/mid-kid/.local/opt/cross/crosstool-ng/.build/arm-none-eabi/build/build-libc/arm-none-eabi/thumb/v8.1-m.main+pacbti+fp/bp/softfp/newlib/targ-include -isystem /home/mid-kid/.local/opt/cross/crosstool-ng/.build/arm-none-eabi/src/newlib/newlib/libc/include -B/home/mid-kid/.local/opt/cross/crosstool-ng/.build/arm-none-eabi/build/build-libc/arm-none-eabi/thumb/v8.1-m.main+pacbti+fp/bp/softfp/libgloss/arm -L/home/mid-kid/.local/opt/cross/crosstool-ng/.build/arm-none-eabi/build/build-libc/arm-none-eabi/thumb/v8.1-m.main+pacbti+fp/bp/softfp/libgloss/libnosys -L/home/mid-kid/.local/opt/cross/crosstool-ng/.build/arm-none-eabi/src/newlib/libgloss/arm  -mthumb -march=armv8.1-m.main+pacbti+fp -mbranch-protection=standard -mfloat-abi=softfp -DPACKAGE_NAME="newlib" -DPACKAGE_TARNAME="newlib" -DPACKAGE_VERSION="4.1.0" -DPACKAGE_STRING="newlib 4.1.0" -DPACKAGE_BUGREPORT="" -DPACKAGE_URL="" -I. -I/home/mid-kid/.local/opt/cross/crosstool-ng/.build/arm-none-eabi/src/newlib/newlib/libm/math -I/home/mid-kid/.local/opt/cross/crosstool-ng/.build/arm-none-eabi/src/newlib/newlib/libm/math/../common -Os -fno-builtin -D__NO_SYSCALLS__ -D_COMPILING_NEWLIB -DHAVE_INIT_FINI      -ftls-model=local-exec  -g -Os  -mthumb -march=armv8.1-m.main+pacbti+fp -mbranch-protection=standard -mfloat-abi=softfp -c -o lib_a-wf_log2.o `test -f 'wf_log2.c' || echo '/home/mid-kid/.local/opt/cross/crosstool-ng/.build/arm-none-eabi/src/newlib/newlib/libm/math/'`wf_log2.c
[ALL  ]    /home/mid-kid/.local/opt/cross/crosstool-ng/.build/arm-none-eabi/src/newlib/newlib/libm/math/wf_log2.c: In function 'log2f':
[ERROR]    /home/mid-kid/.local/opt/cross/crosstool-ng/.build/arm-none-eabi/src/newlib/newlib/libm/math/wf_log2.c:34:22: error: 'float_t' undeclared (first use in this function); did you mean 'float'?
[ALL  ]       34 |   return (logf(x) / (float_t) M_LN2);
[ALL  ]          |                      ^~~~~~~
[ALL  ]          |                      float
[ALL  ]    /home/mid-kid/.local/opt/cross/crosstool-ng/.build/arm-none-eabi/src/newlib/newlib/libm/math/wf_log2.c:34:22: note: each undeclared identifier is reported only once for each function it appears in
[ERROR]    /home/mid-kid/.local/opt/cross/crosstool-ng/.build/arm-none-eabi/src/newlib/newlib/libm/math/wf_log2.c:34:30: error: expected ')' before numeric constant
[ALL  ]       34 |   return (logf(x) / (float_t) M_LN2);
[ALL  ]          |          ~                   ^
[ALL  ]          |                              )
[ERROR]    make[9]: *** [Makefile:1276: lib_a-wf_log2.o] Error 1
```

--- a/packages/newlib/4.1.0/0002-minimal-support-for-ts-18661-3.patch
+++ b/packages/newlib/4.1.0/0002-minimal-support-for-ts-18661-3.patch
@@ -1,0 +1,50 @@
+From 91f99d323b39dc0c06c40038791db9861d4b76b9 Mon Sep 17 00:00:00 2001
+From: Kito Cheng <kito.cheng@sifive.com>
+Date: Mon, 12 Jul 2021 17:17:21 +0800
+Subject: [PATCH] Minimal support for ISO/IEC TS 18661-3.
+
+ - GCC will set __FLT_EVAL_METHOD__ to 16 if __fp16 supported, e.g.
+   cortex-a55/aarch64.
+   - $ aarch64-unknown-elf-gcc -v 2>&1 |grep version
+     gcc version 9.2.0 (GCC)
+   - $ aarch64-unknown-elf-gcc  -E -dM -mcpu=cortex-a55 - < /dev/null  |grep FLT_EVAL_METHOD
+     #define __FLT_EVAL_METHOD__ 16
+     #define __FLT_EVAL_METHOD_TS_18661_3__ 16
+     #define __FLT_EVAL_METHOD_C99__ 16
+ - The behavior of __FLT_EVAL_METHOD__ == 16 is same as
+   __FLT_EVAL_METHOD__ == 0 except for float16_t, but newlib didn't
+   support float16_t.
+
+ISO/IEC TS 18661-3:
+http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2405.pdf
+
+V2 Changes:
+- List Howland, Craig D as co-author since he provide the draft of comment
+  in math.h.
+
+Co-authored-by: "Howland, Craig D" <howland@LGSInnovations.com>
+---
+ newlib/libc/include/math.h |   12 +++++++++++-
+ 1 file changed, 11 insertions(+), 1 deletion(-)
+
+--- a/newlib/libc/include/math.h
++++ b/newlib/libc/include/math.h
+@@ -146,7 +146,17 @@
+   #define __TMP_FLT_EVAL_METHOD
+ #endif /* FLT_EVAL_METHOD */
+ #if defined FLT_EVAL_METHOD
+-  #if FLT_EVAL_METHOD == 0
++/* FLT_EVAL_METHOD == 16 has meaning as defined in ISO/IEC TS 18661-3,
++ * which provides non-compliant extensions to C and POSIX (by adding
++ * additional positive values for FLT_EVAL_METHOD).  It effectively has
++ * same meaning as the C99 and C11 definitions for value 0, while also
++ * serving as a flag that the _Float16 (float16_t) type exists.
++ *
++ * FLT_EVAL_METHOD could be any number of bits of supported floating point
++ * format (e.g. 32, 64, 128), but currently only AArch64 and few other targets
++ * might define that as 16.  */
++  #if (FLT_EVAL_METHOD == 0) \
++      || (FLT_EVAL_METHOD == 16)
+     typedef float  float_t;
+     typedef double double_t;
+    #elif FLT_EVAL_METHOD == 1

--- a/packages/newlib/4.1.0/0003-support-__FLT_EVAL_METHOD__-values.patch
+++ b/packages/newlib/4.1.0/0003-support-__FLT_EVAL_METHOD__-values.patch
@@ -1,0 +1,65 @@
+From 27f0913c17c216548b2f5eea79037ee90038f375 Mon Sep 17 00:00:00 2001
+From: Andrea Corallo <andrea.corallo@arm.com>
+Date: Wed, 30 Mar 2022 15:40:59 +0200
+Subject: [PATCH] Aarch32/64: Support __FLT_EVAL_METHOD__ values other than 0,
+ 1, 2
+
+2022-03-30  Andrea Corallo  <andrea.corallo@arm.com>
+
+	* libc/include/machine/ieeefp.h (__FLOAT_TYPE, __DOUBLE_TYPE): New
+	macros.
+	* libc/include/math.h: Uses __DOUBLE_TYPE __FLOAT_TYPE to define
+	double_t float_t if possible.
+---
+ newlib/libc/include/machine/ieeefp.h |   14 ++++++++++++++
+ newlib/libc/include/math.h           |    9 +++++++++
+ 2 files changed, 23 insertions(+)
+
+--- a/newlib/libc/include/machine/ieeefp.h
++++ b/newlib/libc/include/machine/ieeefp.h
+@@ -90,6 +90,13 @@
+ #ifndef __SOFTFP__
+ # define _SUPPORTS_ERREXCEPT
+ #endif
++/* As per ISO/IEC TS 18661 '__FLT_EVAL_METHOD__' will be defined to 16
++   (if compiling with +fp16 support) so it can't be used by math.h to
++   define float_t and double_t.  For values of '__FLT_EVAL_METHOD__'
++   other than 0, 1, 2 the definition of float_t and double_t is
++   implementation-defined.  */
++#define __DOUBLE_TYPE double
++#define __FLOAT_TYPE float
+ #endif
+ 
+ #if defined (__aarch64__)
+@@ -102,6 +109,13 @@
+ #ifdef __ARM_FP
+ # define _SUPPORTS_ERREXCEPT
+ #endif
++/* As per ISO/IEC TS 18661 '__FLT_EVAL_METHOD__' will be defined to 16
++   (if compiling with +fp16 support) so it can't be used by math.h to
++   define float_t and double_t.  For values of '__FLT_EVAL_METHOD__'
++   other than 0, 1, 2 the definition of float_t and double_t is
++   implementation-defined.  */
++#define __DOUBLE_TYPE double
++#define __FLOAT_TYPE float
+ #endif
+ 
+ #ifdef __epiphany__
+--- a/newlib/libc/include/math.h
++++ b/newlib/libc/include/math.h
+@@ -168,6 +168,15 @@
+    #else
+     /* Implementation-defined.  Assume float_t and double_t have been
+      * defined previously for this configuration (e.g. config.h). */
++
++   /* If __DOUBLE_TYPE is defined (__FLOAT_TYPE is then supposed to be
++      defined as well) float_t and double_t definition is suggested by
++      an arch specific header.  */
++   #ifdef __DOUBLE_TYPE
++    typedef __DOUBLE_TYPE double_t;
++    typedef __FLOAT_TYPE float_t;
++   #endif
++   /* Assume config.h has provided these types.  */
+   #endif
+ #else
+     /* Assume basic definitions.  */

